cohttp-lwt/src/connection_cache.ml

```diff
@@ -0,0 +1,150 @@
+module Make_no_cache (Connection : S.Connection) : sig
+  include S.Connection_cache with module IO = Connection.Net.IO
+
+  val create : ?ctx:Connection.Net.ctx -> unit -> t
+end
+=
+struct
+  module Net = Connection.Net
+  module IO = Net.IO
+  open IO
+
+  type t = ?body:Body.t -> Cohttp.Request.t -> (Cohttp.Response.t * Body.t) IO.t
+
+  let request = Fun.id
+
+  let create ?(ctx = Net.default_ctx) () ?body req =
+    Net.resolve ~ctx (Cohttp.Request.uri req)
+    >>= Connection.connect ~ctx ~persistent:false
+    >>= fun connection ->
+    let res = Connection.request connection ?body req in
+    (* this can be simplified when https://github.com/mirage/ocaml-conduit/pull/319 is released. *)
+    Lwt.async begin fun () ->
+      res >>= fun (_, body) ->
+      match body with
+      | `Empty | `String _ | `Strings _ -> Lwt.return_unit
+      | `Stream stream ->
+        Lwt_stream.closed stream >>= fun () ->
+        Connection.close connection;
+        Lwt.return_unit
+    end;
+    res
+end
+
+module Make (Connection : S.Connection) (Sleep : S.Sleep with type 'a promise = 'a Lwt.t) : sig
+  include S.Connection_cache with module IO = Connection.Net.IO
+
+  val create :
+    ?ctx:Connection.Net.ctx ->
+    ?keep:int64 ->
+    ?retry:int ->
+    ?parallel:int ->
+    ?depth:int ->
+    unit -> t
+end
+=
+struct
+  module Net = Connection.Net
+  module IO = Net.IO
+  open IO
+
+  type ctx = Net.ctx
+
+  exception Retry = S.Retry
+
+  module Endp_map = Map.Make (Net.Endp)
+
+  (* type t = < request : ?body:Body.t -> Cohttp.Request.t -> (Cohttp.Response.t * Body.t) IO.t > *)
+  type t =
+    { mutable cache : Connection.t list Endp_map.t
+    ; ctx : ctx
+    ; keep : int64
+    ; retry : int
+    ; parallel : int
+    ; depth : int
+    }
+
+  (*
+   * Select from existing connections by hostname, not IP.
+   * One hostname may resolve to multiple IP addresses.
+   * In such a case we want only one connection.
+   * The downside is that some resolver redirections may break.
+   *)
+
+  let create ?(ctx = Net.default_ctx) ?(keep=60_000_000_000L) ?(retry=2) ?(parallel=4) ?(depth=100) () =
+    { cache = Endp_map.empty
+    ; ctx ; keep ; retry ; parallel ; depth
+    }
+
+  let rec get_connection self endp =
+    let finalise connection =
+      let conns = Endp_map.find endp self.cache in
+      let conns = List.filter ((!=) connection) conns in
```

Comment:
```ocaml
      let conns = Endp_map.find endp self.cache
                  |> List.filter ((!=) connection)
      in
```
feels more in line with the rest of the style but I don't have a very strong opinion about this